architecture/2024-06-resource-extensibility-feature-spec.md

```diff
@@ -0,0 +1,457 @@
+
+Below is an example extender implementation for PostgreSQL
+
+``` bicep
+extension radius
+
+resource extender 'Applications.Core/extenders@2023-10-01-preview' = {
+  name: 'postgresql'
+  properties: {
+    name : 'postgresql'
+    sku : 'standard'
+    }
+  }
+}
+```
+The properties in the extender resource are free form key value pairs. The user can pass in any text as SKU as there are no validation. In a mature enterprise, platform engineers or IT operators want to control the SKU a developer can pass in and provision. They would allow a set of SKUs preventing developers from choosing expensive instances. User defined types will enable the platform engineer to define the resource type definition in a strong typed way with guardrails and validation rules adhering to their organizational policies. 
+
+## User requirements
+
+As a platform engineer, I want to build a custom platform for application development with the following requirements 
+
+- I want to author the resource schema for a set of key technologies for the enterprise applications. The current extenders in Radius do not provide a way to define custom resource types with strongly typed schema and guardrails.
+- I want to iterate, test, deploy and ship the custom application platform to my IT operators and developers.
+- I want to enable my IT operators to collaborate and maintain the custom resource types within the platform.
+- I want my developers to leverage all the features of Radius such as Recipes, Connections, App graph, rad CLI for custom resource types.
+
+## Proposed solution
+
+## Authoring, Registering and Deploying the user-defined types in Radius
+
+Define a schema for the user-defined type that my application uses in Radius. The schema defines the structure and properties of the custom resource type. The resource schema authoring tool provides guidance on auto-completion to help the user author the schema. 
+
+### Schema authoring format
+
+| Schema Format | Why chose? | Why not to choose? | 
+|---------------|------------|-------------------|
+| OpenAPI YAML/JSON | More common and standard format across the open-source communities ; K8s CRDs follow the structure of Open API  | Managing versions and ensuring backward compatibility is challenging |
+| Typespec | Developer friendly; Maintenance and upgrades are easier; Errors and guidance violations at development time | Microsoft developed language; May not be widely adopted by the cloud-native communities ; Yet another language to adopt |
+
+First, we will create a custom YAML format that requires basic user inputs to define the schema for the user defined type in Radius for users to start using the feature. Following that, we will implement the Typespec authoring experience to provide users with more guidance and validation when authoring the schema.
```

Comment:
I think we should change 'will' implement to 'may' implement Typespec authoring. If we don't get customers asking for it I assume we won't do it...